ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -2178,10 +2178,17 @@ public void replicateWithNetworkClient() {
         }
         List<RequestInfo> requestInfos =
             pollRemoteReplicaGroups(remoteReplicaGroups, correlationIdToRequestInfo, correlationIdToReplicaGroup);
-        Set<Integer> requestsToDrop = filterTimedOutRequests(correlationIdToRequestInfo, correlationIdToReplicaGroup);
+        List<ResponseInfo> responseInfosForTimedOutRequests =
+            filterTimedOutRequests(correlationIdToRequestInfo, correlationIdToReplicaGroup);
+        Set<Integer> requestsToDrop = responseInfosForTimedOutRequests.stream()
+            .map(r -> r.getRequestInfo().getRequest().getCorrelationId())
+            .collect(Collectors.toSet());
 
         final int pollTimeoutMs = (int) replicationConfig.replicationRequestNetworkPollTimeoutMs;
         List<ResponseInfo> responseInfos = networkClient.sendAndPoll(requestInfos, requestsToDrop, pollTimeoutMs);
+        // Add response for dropped request because there is no guarantee that sendAndPoll would return a response for
+        // dropped requests.
+        responseInfos.addAll(responseInfosForTimedOutRequests);
```

Comment:
What if the sendAndPoll returns response for the dropped requests? We may have two responses for the single requests. 
I guess we reply on the factor that inside the onResponses function, if the correlationId doesn't exist, we do nothing. 
Can we add comment for it?

Reply:
Done.